### PR TITLE
Simplify seed creation/validation commands

### PR DESCRIPTION
**Pull request by goodov**
Removed unnecessary commands, reworked validation logic to always return errors as array of strings instead of throwing.

### CODE REVIEW

**Comment by goodov** [comment]
@brave/griffin-maintainers ptal

File: src/seed_tools/commands/create_seed.ts [hunk 1]
```diff
@@ -13,69 +13,165 @@ import {
...
+      'output seed file, may be omitted if --validate_only is provided',
+    )
+    .option('--fix', 'fix format errors in-place')
+    .option('--mock_serial_number <value>', 'mock serial number')
+    .option(
+      '--output_serial_number_file <path>',
+      'file path to write the seed serial number',
+      './serialnumber',
+    )
+    .option('--validate_only', 'validate the seed without creating it')
```
Comment (atuchin-m) [inline src/seed_tools/commands/create_seed.ts:36]: The name sounds confusing to me with a combination of `--fix`.
What if we just remove it and use `output_seed_file == null` instead?

File: src/seed_tools/commands/create_seed.ts [hunk 2]
```diff
@@ -13,69 +13,165 @@ import {
...
-    if (!dirItem.isFile()) {
-      continue;
-    }
-    files.push(dirItem.name);
+async function main(
+  studiesDir: string,
+  outputSeedFile: string | undefined,
+  options: Options,
+) {
+  console.log(outputSeedFile, options);
```
Comment (atuchin-m) [inline src/seed_tools/commands/create_seed.ts:54]: do we need this? if yes, let's add move meaningful description what it prints.

File: src/seed_tools/commands/create_seed.ts [hunk 3]
```diff
@@ -13,69 +13,165 @@ import {
...
+  if (outputSeedFile === undefined && !options.validate_only) {
+    console.error(
+      'Either output_seed_file or --validate_only must be provided',
+    );
+    process.exit(1);
   }
-  files.sort();
+
+  if (options.fix && outputSeedFile) {
```
Comment (atuchin-m) [inline src/seed_tools/commands/create_seed.ts:62]: nit: in the other places we use comparing with `undefined`. Let's unify this.

File: src/seed_tools/utils/study_json_utils.ts [hunk 4]
```diff
@@ -13,42 +13,36 @@ export interface Options {
...
-}
-
-export async function readStudyFileReturnWithError(
-  studyFilePath: string,
-  options?: Options,
-): Promise<[Study[], string] | Error> {
+): Promise<{
+  studies: Study[];
+  studyFileContent: string;
+  errors: string[];
```
Comment (atuchin-m) [inline src/seed_tools/utils/study_json_utils.ts:19]: nit: `error?: string;`

File: src/seed_tools/utils/study_validation.test.ts [hunk 5]
```diff
@@ -383,12 +419,16 @@ describe('validateStudy', () => {
       ],
     });
 
-    expect(() => {
-      study_validation.validateStudy(study, studyFilePath);
-    }).toThrowError('Duplicate param name: test in experiment experiment1');
+    expect(
```
Comment (atuchin-m) [inline src/seed_tools/utils/study_validation.test.ts:422]: any options to implement a helper like `expectHasError(study_validation.validateStudy(study, studyFilePath), 'error-string')`?

File: src/seed_tools/commands/create_seed.ts [hunk 6]
```diff
@@ -13,69 +13,165 @@ import {
...
     const filePath = path.join(studiesDir, file);
-    const studies = await study_json_utils.readStudyFile(filePath);
-    for (const study of studies) {
-      study_validation.validateStudy(study, filePath);
-      setStudyFixedParameters(study);
-      variationsSeed.study.push(study);
+    const readStudyFileResult = await study_json_utils.readStudyFile(filePath);
+    errors.push(...readStudyFileResult.errors);
+    errors.push(
+      ...(await checkAndOptionallyFixFormat(
```
Comment (atuchin-m) [inline src/seed_tools/commands/create_seed.ts:122]: if there are `readStudyFileResult.errors` this code incorrectly suggests to "format" the file by replacing it to `[]` ;) 
```
Found unknown field while reading variations.Study.Experiment.Param from JSON format. JSON key: name1 (studies/BraveAdsTextClassificationPageProbabilitiesStudy.json)
---
Format required:
diff --git studies/BraveAdsTextClassificationPageProbabilitiesStudy.json studies/BraveAdsTextClassificationPageProbabilitiesStudy.json.formatted
index 5055002..fe51488 100644
--- studies/BraveAdsTextClassificationPageProbabilitiesStudy.json
+++ studies/BraveAdsTextClassificationPageProbabilitiesStudy.json.formatted
@@ -1,40 +1 @@
-[
-  {
-    "name": "BraveAdsTextClassificationPageProbabilitiesStudy",
...
-  }
-]
+[]
```

**Comment by atuchin-m** [comment]
BTW, we don't have a protection against duplicating some things:
```
      "channel": [
        "NIGHTLY",
        "BETA",
        "RELEASE",
         "NIGHTLY" // <= a possbile issue here
      ],
 ```
 
 This is also considered only as a format error:
 ```
     "experiment": [
      {
        "name": "Enabled",
        "name": "AnotherEnabled",
        ...
        
```

File: src/seed_tools/commands/create_seed.ts (hunk 1 above)
Comment (goodov) [reply src/seed_tools/commands/create_seed.ts]: fixed by introducing separate `create` and `lint` commands.

File: src/seed_tools/utils/study_validation.test.ts (hunk 5 above)
Comment (goodov) [reply src/seed_tools/utils/study_validation.test.ts:442]: will be replaced later with tests cleanup.

File: src/seed_tools/commands/create_seed.ts (hunk 6 above)
Comment (goodov) [reply src/seed_tools/commands/create_seed.ts]: thanks, fixed.

**Review by atuchin-m — APPROVED** [review]

File: src/seed_tools/utils/studies_to_seed.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  errors: string[];
+}> {
+  const files = (await fs.readdir(studiesDir)).sort();
+
+  const studies: Study[] = [];
+  const studyFileBaseNameMap = new Map<Study, string>();
+  const errors: string[] = [];
+
+  for (const file of files) {
+    const filePath = path.join(studiesDir, file);
```
[the remark this answers is not in the document]
Comment (kdenhartog) [reply src/seed_tools/utils/studies_to_seed.ts:56]: I'm not seeing anyway that someone could produce a file traversal vulnerability with this. readdir returns an array of the current files in the directory and excludes the usage of `'.' and '..'`. This seems fine to me.

**Review by kdenhartog — APPROVED** [review]

**Pull request merged**